### PR TITLE
Implement AppDynamics plugin

### PR DESCRIPTION
**Pull request by mohabusama**
+ Healthrule violations check
+ Filter by severity

API used: https://docs.appdynamics.com/display/PRO41/Use+the+AppDynamics+REST+API#UsetheAppDynamicsRESTAPI-Retrieveallhealthruleviolationsinaparticularbusinessapplication

Not sure if we will need to implement Events API as well:
https://docs.appdynamics.com/display/PRO41/Use+the+AppDynamics+REST+API#UsetheAppDynamicsRESTAPI-Retrieveeventdata

### CODE REVIEW

File: zmon_worker_monitor/builtins/plugins/appdynamics.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        """
+        Automatically called to create the check function's object
+        :param factory_ctx: (dict) names available for Function instantiation
+        :return: an object that implements a check function
+        """
+        return propartial(AppdynamicsWrapper, username=self._user, password=self._pass)
+
+
+class AppdynamicsWrapper(object):
+    def __init__(self, url, username=None, password=None):
```
Comment (Jan-M) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: This needs to be changed to support using oauth2.

We will not share credentials with zmon directly but use skoap/skipper as a oauth2 secured proxy so only we know the credentials.

Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: Does this mean we will support oauth2 only or credentials as well?

Comment (Jan-M) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: We can continue to have the credentials implemented, but for Z it will never be used imho.

Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: Done. Tested with credentials on ZMon vagrant env.

**Comment by Jan-M** [comment]
:+1:

File: zmon_worker_monitor/builtins/plugins/appdynamics.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            self._session.headers.update({'Authorization': 'Bearer {}'.format(tokens.get('uid'))})
+        else:
+            self._session.auth = (username, password)
+
+        self._session.headers.update({'User-Agent': get_user_agent()})
+        self._session.params = {'output': 'json'}
+        self._session.timeout = 3
+
+    @property
+    def session(self):
```
Comment (hjacobs) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: this should be hidden, i.e. not accessible from ZMON check command code

make it start with two underscores

File: zmon_worker_monitor/builtins/plugins/appdynamics.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        """
+        return propartial(AppdynamicsWrapper, username=self._user, password=self._pass)
+
+
+class AppdynamicsWrapper(object):
+    def __init__(self, url, username=None, password=None):
+        self.url = url
+        self.timeout = 3
+
+        self._session = requests.Session()
```
Comment (hjacobs) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: this should be hidden, i.e. not accessible from ZMON check command code

make it start with two underscores

Comment (mohabusama) [inline zmon_worker_monitor/builtins/plugins/appdynamics.py]: Done with update code safety check.

**Comment by hjacobs** [comment]
:+1:

**Comment by hjacobs** [comment]
Hello Zappr?

**Comment by mohabusama** [comment]
👍

**Comment by hjacobs** [comment]
:+1:

**Pull request merged**